proposed/2021/ProjectUpdateEventsInVisualStudio.md

```diff
@@ -0,0 +1,152 @@
+        /// The list will not include all projects. Some projects may have been skipped in earlier up to date check, and other projects may no-op.
+        /// </summary>
+        /// <remarks>
+        /// Just because a project is being restored that doesn't necessarily mean any actual updates will happen.
+        /// Only PackageReference projects are includede in this list.
+        /// </remarks>
+        event SolutionRestoreEventHandler SolutionRestoreFinished;
+
+        /// <summary>
+        /// Raised when particular project is about to be updated.
+        /// For PackageReference projects, this means an assets file or a nuget temp msbuild file write (nuget.g.props or nuget.g.targets). The list of updated files will include the aforementioned.
+        /// For packages.config projects, this means a single package is installed/unistall/unistalled. The list of updated files may include the path of the package that was changed.
+        /// </summary>
+        event ProjectUpdateEventHandler ProjectUpdateStarted;
+
+        /// <summary>
+        /// Raised when particular project update has been completed.
+        /// For PackageReference projects, this means an assets file or a nuget temp msbuild file write (nuget.g.props or nuget.g.targets). The list of updated files will include the aforementioned.
+        /// For packages.config projects, this means a single package is installed/unistall/unistalled. The list of updated files may include the path of the package that was changed.
+        /// </summary>
+        event ProjectUpdateEventHandler ProjectUpdateFinished;
+    }
+
+    /// <summary>
+    /// Defines an event handler delegate for PackageReference solution restore start and end.
+    /// </summary>
+    /// <param name="projects">List of projects that will run restore. Never <see langword="null"/>.</param>
+    public delegate void SolutionRestoreEventHandler(IReadOnlyList<string> projects);
+
+    /// <summary>
+    /// Defines an event handler delegate for project updates.
+    /// </summary>
+    /// <param name="projectUniqueName">Project full path. Never <see langword="null"/>. </param>
+    /// <param name="updatedFiles">NuGet output files that may be updated. Never <see langword="null"/>.</param>
+    public delegate void ProjectUpdateEventHandler(string projectUniqueName, IReadOnlyList<string> updatedFiles);
+}
+```
+
+A few other notes:
+
+- `SolutionRestoreStarted` and `SolutionRestoreFinished` events will only be called when a solution restore happens, as a response to a solution/project load, or a user issued restore. Package installation through PM UI, PMC or any of the NuGet extensibility APIs *will never* call this event.
+- `ProjectUpdateStarted` and `ProjectUpdateFinished` may be called without corresponding `SolutionRestoreStarted` and `SolutionRestoreFinished` events, if a package installation has been triggered through PM UI, PMC or any of the NuGet extensibility APIs.
```

Comment:
But PM UI or PMC triggers a restore _after_ a package is modifed (installed, uninstalled, updated)

In that case, will we have 

- ProjectUpdateStarted: From package install
- ProjectUpdateFinished: From package install
- SolutionRestoreStarted: From PMUI/PMC restore action
- SolutionRestoreFinished: From PMUI/PMC restore action

?

Reply:
The restore in PackageReference following a package change through PMC or PM UI will be no-op.